content/client-lib-development-guide/features.textile

```diff
@@ -468,7 +468,7 @@ h3(#realtime-presence). Presence
 * @(RTP1)@ When a channel @ATTACHED@ @ProtocolMessage@ is received, the @ProtocolMessage@ may contain a @HAS_PRESENCE@ bit flag indicating that there are currently members present on the channel, see "TR3":#TR3 . If the flag is 1, the server will shortly perform a @SYNC@ operation as described in "RTP18":#RTP18 . If that flag is 0 or there is no @flags@ field, the presence map should be considered in sync immediately with no members present on the channel
 * @(RTP2)@ A @PresenceMap@ should be used to maintain a list of members present on a channel. Broadly, this is is a map of "memberKeys":#TP3h to presence messages, all with @PRESENT@ actions (during a sync there may also be ones with an @ABSENT@ action, see "RTP2f":#RTP2f ). Consult the "the Java implementation":https://github.com/ably/ably-java/blob/master/lib/src/main/java/io/ably/lib/realtime/Presence.java#L578-L723 for details.
 ** @(RTP2a)@ All incoming presence messages must be compared for newness with the matching member already in the @PresenceMap@, if one exists, where "matching" means they share the same @memberKey@ (or equivalently, they share both @connectionId@ and @clientId@)
-** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of an incoming presence message corresponds to its @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
+** @(RTP2b)@ 'Comparing for newness' involves: If the @id@ of both presence message corresponds to the @connectionId@ (defined as the @connectionId@ being an initial substring of the @id@), then the messages are compared lexicographically by @id@. If not, they are compared by timestamp
```

Comment:
> After thinking about it, I think we should go with i.

So, given that:

- we're going for minimum viable change in 0.9 to get the new features we want;

- any logical clock proposal, whilst preferable longer term for various reasons, won't be fully thought-through if we do it immediately, and it'll therefore be wrong.

I agree.

Although I don't like having to do anything non-essential on a per-message basis, I think this is ok; a simple optimisation a library can do is to compare each msg id to verify that they start with the connectionId; if they do, and are the same length, then a lexicographic comparison does the job. If they are different lengths then more work is needed (actually if we guarantee that the string sent by realtime never has leading zeros in the msgSerial part of the id, then a longer msg id is later than a shorter msg id.